src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,223 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/core/CHIPTLV.h"
+#include "lib/core/CHIPTLVTags.h"
+#include "lib/core/CHIPTLVTypes.h"
+#include "protocols/interaction_model/Constants.h"
+#include "system/SystemPacketBuffer.h"
+#include "system/TLVPacketBufferBackingStore.h"
+#include <app/BufferedReadCallback.h>
+#include <app/InteractionModelEngine.h>
+
+namespace chip {
+namespace app {
+
+void BufferedReadCallback::OnReportBegin()
+{
+    mCallback.OnReportBegin();
+}
+
+void BufferedReadCallback::OnReportEnd()
+{
+    DispatchBufferedData(nullptr, mBufferedPath, StatusIB(), true);
+    mCallback.OnReportEnd();
+}
+
+CHIP_ERROR BufferedReadCallback::GenerateListTLV(System::PacketBufferTLVReader & aReader)
+{
+    System::PacketBufferHandle handle = System::PacketBufferHandle::New(1000);
+    System::PacketBufferTLVWriter writer;
+    CHIP_ERROR err;
+    TLV::TLVType outerType;
+
+    //
+    // Initialize the writer, but with chaining enabled, since we'll very likely
+    // exceed the size of a single packet buffer when reconstituing the entire list
+    //
+    writer.Init(std::move(handle), /* useChainedBuffers */ true);
+
+    err = writer.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Array, outerType);
+    SuccessOrExit(err);
+
+    for (auto & bufHandle : mBufferedList)
+    {
+        System::PacketBufferTLVReader reader;
+
+        reader.Init(std::move(bufHandle));
+
+        err = reader.Next();
+        SuccessOrExit(err);
+
+        err = writer.CopyElement(TLV::AnonymousTag, reader);
+        SuccessOrExit(err);
+    }
+
+    err = writer.EndContainer(outerType);
+    SuccessOrExit(err);
+
+    writer.Finalize(&handle);
+    aReader.Init(std::move(handle));
+
+exit:
+    return err;
+}
+
+CHIP_ERROR BufferedReadCallback::BufferData(const ConcreteAttributePath & aPath, TLV::TLVReader * apData)
+{
+    CHIP_ERROR err;
+
+    if (aPath.mListOp == ConcreteAttributePath::ListOperation::ReplaceAll)
+    {
+        TLV::TLVType outerContainer;
+
+        VerifyOrExit(apData->GetType() == TLV::kTLVType_Array, err = CHIP_ERROR_INVALID_TLV_ELEMENT);
+        mBufferedList.clear();
+
+        err = apData->EnterContainer(outerContainer);
+        SuccessOrExit(err);
+
+        while ((err = apData->Next()) == CHIP_NO_ERROR)
+        {
+            System::PacketBufferTLVWriter writer;
+            System::PacketBufferHandle handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+            writer.Init(std::move(handle), false);
+
+            err = writer.CopyElement(TLV::AnonymousTag, *apData);
+            SuccessOrExit(err);
+
+            err = writer.Finalize(&handle);
```

Comment:
Does Finalize have any provision to return unused portions of the chain buffer to the system?

If not, I worry that this code is reliant upon peers sending large chunks.  But to my knowledge, that is actually not required.  And if peers send a large number of very small chunks, can't the recipient end up with a very large number of kMaxSecureSduLengthBytes-size allocations in mBufferedList?

It seems like it could be important to use remaining capacity in the most recent, previously allocated buffer in mBufferedList if any exists.  Thoughts?

Reply:
Good catch.

`PacketBufferHandle` provides a method called `RightSize()` that on supported platforms, will allocate a new more shrink-wrapped packet buffer and copy the contents over to that. This will ensure no wasted space for each list element stored within the buffered list. Will update the code to make that call.

The current design however, takes the path of providing a separate packet buffer for each list element stored in `mBufferedList`. If you wanted to be able to use a single, (chained) packet buffer to back all list items stored in the buffered list, that actually ends up being quite problematic. This is certainly ok if we end up receiving each list item in order, but if we end up needing to 'patch up' a list item out of order (e.g if we get an update operation), and the list item grows in size, then it needs to _push out_ the other list items below it in the buffer, which is quite complicated logic.

So instead, I chose the path of expedience here.